src/vaults/Distributor.ts

```diff
@@ -0,0 +1,228 @@
+      }, constants.Zero);
+
+      console.log(
+        `Total MORPHO emitted: ${formatUnits(totalTokenEmitted)} over ${formatUnits(
+          this._morphoAccumulatedFromMainDistribution
+        )} MORPHO available`
+      );
+
+      console.log(`Total MORPHO distributed during ${epochConfig.id}: ${formatUnits(lastEpochDistributed)}`);
+      lastEpochDistributed = totalTokenEmitted;
+      // process of the distribution and the merkle tree
+      trees[epochConfig.id] = this._computeCurrentMerkleTree();
+    }
+
+    const lastEpochId = epochsProofs[epochsProofs.length - 1].epoch;
+    return {
+      lastMerkleTree: trees[lastEpochId],
+      history: trees,
+    };
+  }
+
+  private _updateMarketIndex(rate: BigNumber, timestampTo: BigNumber) {
+    const timeDelta = timestampTo.sub(this._lastTimestamp);
+    const morphoAccrued = timeDelta.mul(rate);
+    if (this._totalSupply.gt(0))
+      this._marketIndex = this._marketIndex.add(WadRayMath.wadDiv(morphoAccrued, this._totalSupply));
+    this._lastTimestamp = BigNumber.from(timestampTo);
+    return morphoAccrued;
+  }
+
+  private _processEndOfEpoch(rate: BigNumber, epochConfig: EpochConfig) {
+    if (this._totalSupply.gt(0)) this._updateMarketIndex(rate, epochConfig.finalTimestamp);
+
+    Object.values(this._usersConfigs).forEach((userConfig) => {
+      if (!userConfig) return;
+      userConfig.morphoAccrued = userConfig.morphoAccrued.add(
+        WadRayMath.wadMul(this._marketIndex.sub(userConfig.index), userConfig.balance).div(Distributor.SCALING_FACTOR)
+      );
+      userConfig.index = this._marketIndex;
+    });
+  }
+
+  private async _handleTransaction(transaction: TransactionEvents, rate: BigNumber) {
+    // process event
+    // we first update the global vault distribution
+    const block = await this.eventsFetcher.getBlock(transaction.event.blockNumber);
+    this._updateMarketIndex(rate, BigNumber.from(block.timestamp));
+
+    // and then distribute to the user(s) of the transaction
+    switch (transaction.type) {
+      case VaultEventType.Deposit: {
+        const event = transaction.event as DepositEvent;
+        this._handleDeposit(event);
+        break;
+      }
+      case VaultEventType.Withdraw: {
+        const event = transaction.event as WithdrawEvent;
+        this._handleWithdrawEvent(event);
+        break;
+      }
+      case VaultEventType.Transfer: {
+        const event = transaction.event as TransferEvent;
+        this._handleTransferEvent(event);
+        break;
+      }
+    }
+  }
+
+  private _handleDeposit(event: DepositEvent) {
+    const userBalance = this._getUserConfig(event.args.owner);
+    userBalance.morphoAccrued = userBalance.morphoAccrued.add(
+      WadRayMath.wadMul(this._marketIndex.sub(userBalance.index), userBalance.balance).div(Distributor.SCALING_FACTOR)
+    );
+    userBalance.balance = userBalance.balance.add(event.args.shares);
+    userBalance.index = BigNumber.from(this._marketIndex);
```

Comment:
this could be abstract away as a `_increaseUserBalance` method as it is used twice (once here, and a second time in transfer)